### PR TITLE
H5 md serial output on head parallel

### PR DESCRIPTION
**Pull request by jonaslandsgesell**
I made H5MD output standard conform (the dimensionality of the datasets that were written was in some cases wrong). Further I added fast ordered output for h5md from the master node only. In the case of having tons of particles the output might be faster with the option write_ordered=false which performs writing in parallel on all nodes, however data is not sorted according to ids.

### CODE REVIEW

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 1]
```diff
@@ -82,17 +116,24 @@ void File::InitFile()
     H5Eset_auto(H5E_DEFAULT, (H5E_auto_t) H5Eprint, stderr);
     std::cout << "Called " << __func__ << " on node " << this_node << std::endl;
 #endif
+    //use a seperate mpi communicator if we want to write out ordered data. This is in order to avoid  blocking by collective functions
+    if(m_write_ordered ==true)
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 2]
```diff
@@ -82,17 +116,24 @@ void File::InitFile()
...
+        MPI_Comm_split(MPI_COMM_WORLD, this_node, 0, &MPI_H5MD_COMM); 
+    else
+        MPI_H5MD_COMM=MPI_COMM_WORLD;
+    if(m_write_ordered== true and this_node!=0)
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 3]
```diff
@@ -82,17 +116,24 @@ void File::InitFile()
...
+    if(n_part <= 0) {
         throw std::runtime_error("Please first set up particles before initializing the H5md object.");
     }
-
     init_filestructure();
     bool fileexists = check_file_exists(m_filename);
     /* Perform a barrier synchronization. Otherwise one process might already
      * create the file while another still checks for its existence. */
-    MPI_Barrier(MPI_COMM_WORLD);
+    if(m_write_ordered==false)
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 4]
```diff
@@ -199,17 +240,19 @@ void File::create_datasets(bool only_load)
                 datasets[path] = h5xx::dataset(groups[father],
                                                basename);
             } else {
-                auto dims = create_dims(descr.dim, descr.size);
-                auto cdims = create_dims(descr.dim, descr.size, 1);
-                auto maxdims = create_maxdims(descr.dim, descr.size);
-                auto storage = h5xx::policy::storage::chunked(cdims).set(
+                int creation_size_dataset=0;
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 5]
```diff
@@ -199,17 +240,19 @@ void File::create_datasets(bool only_load)
...
+                int chunk_size=1;
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 6]
```diff
@@ -199,17 +240,19 @@ void File::create_datasets(bool only_load)
...
+                if(descr.dim>1){
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 7]
```diff
@@ -199,17 +240,19 @@ void File::create_datasets(bool only_load)
...
+                    //we deal now with a particle based property
+                    chunk_size=n_part;
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 8]
```diff
@@ -236,15 +280,19 @@ void File::create_new_file(const std::string &filename)
...
 
     create_groups();
-    create_datasets(false);
+    bool only_load = false;
+    create_datasets(only_load);
     std::vector<double> boxvec = {box_l[0], box_l[1], box_l[2]};
     h5xx::write_attribute(groups["particles/atoms/box"], "dimension", 3);
     h5xx::write_attribute(groups["particles/atoms/box"], "boundary", "periodic");
-    h5xx::write_dataset(datasets["particles/atoms/box/edges"], boxvec);
+    std::string path_edges="particles/atoms/box/edges";
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 9]
```diff
@@ -257,79 +305,99 @@ void File::Close()
...
-             local_part_id < local_cell->n; ++local_part_id)
-        {
-            const auto & current_particle = local_cell->part[local_part_id];
-            id[0][particle_index][0] = current_particle.p.identity;
+	        id[0][particle_index][0] = current_particle->p.identity;
             if (write_typ)
-            {
-                typ[0][particle_index][0] = current_particle.p.type;
-            }
+                typ[0][particle_index][0] =current_particle->p.type;
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 10]
```diff
@@ -257,79 +305,99 @@ void File::Close()
...
+
+}
+
+void File::Write(int write_dat)
+{
+#ifdef H5MD_DEBUG
+    std::cout << "Called " << __func__ << " on node " << this_node << std::endl;
+#endif
+    int num_particles_to_be_written;
+    if(m_write_ordered==true && this_node==0)
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blanks

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 11]
```diff
@@ -257,79 +305,99 @@ void File::Close()
...
+        num_particles_to_be_written=n_part;
+    else if(m_write_ordered == true && this_node !=0)
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 12]
```diff
@@ -257,79 +305,99 @@ void File::Close()
...
+        return ;
+    else if (m_write_ordered ==false)
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 13]
```diff
@@ -257,79 +305,99 @@ void File::Close()
...
+        num_particles_to_be_written=cells_get_n_particles();
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 14]
```diff
@@ -257,79 +305,99 @@ void File::Close()
...
+
+    double_array_3d pos(boost::extents[1][num_particles_to_be_written][3]);
+    double_array_3d vel(boost::extents[1][num_particles_to_be_written][3]);
+    double_array_3d f(boost::extents[1][num_particles_to_be_written][3]);
+    int_array_3d image(boost::extents[1][num_particles_to_be_written][3]);
+    int_array_3d id(boost::extents[1][num_particles_to_be_written][1]);
+    int_array_3d typ(boost::extents[1][num_particles_to_be_written][1]);
+    double_array_3d mass(boost::extents[1][num_particles_to_be_written][1]);
+
+    if (m_write_ordered==true) {
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 15]
```diff
@@ -257,79 +305,99 @@ void File::Close()
...
+        if( this_node ==0){
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blank

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 16]
```diff
@@ -257,79 +305,99 @@ void File::Close()
...
+    double_array_3d f(boost::extents[1][num_particles_to_be_written][3]);
+    int_array_3d image(boost::extents[1][num_particles_to_be_written][3]);
+    int_array_3d id(boost::extents[1][num_particles_to_be_written][1]);
+    int_array_3d typ(boost::extents[1][num_particles_to_be_written][1]);
+    double_array_3d mass(boost::extents[1][num_particles_to_be_written][1]);
+
+    if (m_write_ordered==true) {
+        if( this_node ==0){ 
+            //loop over all particles
+            for(int particle_index=0;particle_index<n_part;particle_index++){
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: blanks

File: src/core/io/writer/h5md/h5md_core.cpp [hunk 17]
```diff
@@ -367,80 +462,79 @@ void File::WriteDataset(T &data, const std::string& path)
...
     /* Only master node writes time and timestep */
     if (this_node == 0)
-        timecount[0] = timecount[1] = timecount[2] = 1;
-    dims[0] += 1;
-    H5Dset_extent(time.hid(), dims);
-
-    ds = H5Dget_space(time.hid());
-    H5Sselect_hyperslab(ds, H5S_SELECT_SET, timeoffset, NULL, timecount, NULL);
-    ds_new = H5Screate_simple(3, timecount, maxdims);
+        time_count[0] = 1; //write 1 timestep
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.cpp]: indentation

File: src/core/io/writer/h5md/h5md_core.hpp [hunk 18]
```diff
@@ -94,8 +95,11 @@ class File
         std::string &scriptname() { return m_scriptname; };
         // Returns the int that describes which data should be written to the dataset.
         int &what() { return m_what; };
+        // Returns the boolean value that describes whether data should be written to the dataset in the order of ids (possibly slower on output for many particles).
+        bool &write_ordered() {return m_write_ordered;} ;
 
     private:
+    	MPI_Comm MPI_H5MD_COMM;
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.hpp]: indentation

File: src/core/io/writer/h5md/h5md_core.hpp [hunk 19]
```diff
@@ -107,13 +111,28 @@ class File
          * @param filename The Name of the hdf5-file to check.
          * @return TRUE if H5MD structure is present, FALSE else.
          */
-        bool check_for_H5MD_structure(std::string const &filename);
+        bool check_for_H5MD_structure(std::string const &filename);
```
Comment (KaiSzuttor) [inline src/core/io/writer/h5md/h5md_core.hpp]: remove blanks at end of line

File: src/python/espressomd/io/writer/h5md.py [hunk 20]
```diff
@@ -30,8 +30,10 @@ def __init__(self, **kwargs):
                      If types should be written.
         write_mass : bool, optional
                      If masses should be written.
+        write_ordered: bool, optional
+                       If particle properties should be ordered according to ids.
         """
-        self.valid_params = ['filename',]
+        self.valid_params = ['filename','write_ordered',]
```
Comment (KaiSzuttor) [inline src/python/espressomd/io/writer/h5md.py]: blank after comma

File: src/python/espressomd/CMakeLists.txt [hunk 21]
```diff
@@ -91,6 +91,7 @@ foreach(cython_file ${cython_SRC})
                          -I ${CMAKE_CURRENT_SOURCE_DIR}/_espresso
                          ${cython_file}
                          -o ${outputpath}
+                         WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/..
```
Comment (KaiSzuttor) [inline src/python/espressomd/CMakeLists.txt]: this is already included in the espressomd/python branch

File: src/python/espressomd/io/writer/h5md.py [hunk 22]
```diff
@@ -50,9 +52,13 @@ def __init__(self, **kwargs):
                     raise ValueError("{} has to be a bool value.".format(i))
             elif i not in self.valid_params:
                 raise ValueError("Unknown parameter {} for H5MD writer.".format(i))
+        
+        write_ordered_default=False
+        self.write_ordered=kwargs.get('write_ordered',write_ordered_default)
```
Comment (KaiSzuttor) [inline src/python/espressomd/io/writer/h5md.py]: blank after comma

File: src/core/io/writer/h5md/h5md_core.hpp (hunk 18 above)
Comment (fweik) [inline src/core/io/writer/h5md/h5md_core.hpp]: The variable should be called m_comm or something like that. See https://github.com/espressomd/espresso/wiki/Naming-conventions-and-folder-structure

**Comment by fweik** [comment]
Looks good to me, @KaiSzuttor  what do you think?

**Comment by KaiSzuttor** [comment]
i really like the blanks.

Looks good, the documentation should be updated before merge.

**Comment by fweik** [comment]
LGTM

**Pull request merged**